nemo/collections/asr/models/k2_sequence_models.py

```diff
@@ -0,0 +1,393 @@
+        """
+        Changes vocabulary of the tokenizer used during CTC decoding process.
+        Use this method when fine-tuning on from pre-trained model.
+        This method changes only decoder and leaves encoder and pre-processing modules unchanged. For example, you would
+        use it if you want to use pretrained encoder when fine-tuning on a data in another language, or when you'd need
+        model to learn capitalization, punctuation and/or special characters.
+
+        Args:
+            new_tokenizer_dir: Path to the new tokenizer directory.
+            new_tokenizer_type: Either `bpe` or `wpe`. `bpe` is used for SentencePiece tokenizers,
+                whereas `wpe` is used for `BertTokenizer`.
+
+        Returns: None
+
+        """
+        super().change_vocabulary(new_tokenizer_dir, new_tokenizer_type)
+
+        loss_kwargs = self._cfg.get("loss", {})
+
+        if self.use_graph_lm:
+            self.token_lm = None
+            self.token_lm_cache_dict = None
+            self.token_lm_path = None
+            logging.warning(
+                f"""With .change_vocabulary() call for a model with criterion_type=`{self.loss.criterion_type}`,
+                            either a new token_lm or a token_lm_path has to be set manually."""
+            )
+
+        self._update_k2_modules(loss_kwargs)
+
+    @typecheck()
+    def forward(
+        self, input_signal=None, input_signal_length=None, processed_signal=None, processed_signal_length=None,
+    ):
+        """
+        Forward pass of the model.
+
+        Args:
+            input_signal: Tensor that represents a batch of raw audio signals,
+                of shape [B, T]. T here represents timesteps, with 1 second of audio represented as
+                `self.sample_rate` number of floating point values.
+            input_signal_length: Vector of length B, that contains the individual lengths of the audio
+                sequences.
+            processed_signal: Tensor that represents a batch of processed audio signals,
+                of shape (B, D, T) that has undergone processing via some DALI preprocessor.
+            processed_signal_length: Vector of length B, that contains the individual lengths of the
+                processed audio sequences.
+
+        Returns:
+            A tuple of 3 elements -
+            1) The log probabilities tensor of shape [B, T, D].
+            2) The lengths of the acoustic sequence after propagation through the encoder, of shape [B].
+            3) The greedy token predictions of the model of shape [B, T] (via argmax)
+        """
+        # trying to load token_lm from token_lm_cache_dict or token_lm_path if it hasn't been loaded yet
+        if self.use_graph_lm and self.token_lm is None:
+            if self.token_lm_cache_dict is not None:
+                logging.info(f"""Loading token_lm from the dict cache at the first .forward() call.""")
```

Comment:
Since so much code is same between the two move it into a utility function of the Mixin and call it in both forwards.

Reply:
Done in [3ed4f2b](https://github.com/NVIDIA/NeMo/pull/3351/commits/3ed4f2b9c4dfd9359d92b0771ef010c1bab380c7)